src/udisks2blockdevices.cpp

```diff
@@ -51,62 +51,72 @@ BlockDevices *BlockDevices::instance()
 
 BlockDevices::~BlockDevices()
 {
-    qDeleteAll(m_blockDevices);
+    qDeleteAll(m_activeBlockDevices);
     qDeleteAll(m_partitionWaits);
 
-    m_blockDevices.clear();
+    m_activeBlockDevices.clear();
     m_partitionWaits.clear();
     sharedInstance = nullptr;
 }
 
 bool BlockDevices::contains(const QString &dbusObjectPath) const
 {
-    return m_blockDevices.contains(dbusObjectPath);
+    return m_activeBlockDevices.contains(dbusObjectPath);
 }
 
 void BlockDevices::remove(const QString &dbusObjectPath)
 {
     if (contains(dbusObjectPath)) {
-        Block *block = m_blockDevices.take(dbusObjectPath);
+        Block *block = m_activeBlockDevices.take(dbusObjectPath);
+        m_blockDevices.remove(dbusObjectPath);
```

Comment:
Is it possible that the block device was not in `m_activeBlockDevices` but it is in `m_blockDevices`? And if so should it be removed if that happens?

From what I could gather is that these would be first added to `m_blockDevices` and then to `m_activeBlockDevices` which means the answer to the first question could be yes.

Also should we care about `m_pendingBlockDevices` here? I think those might be left behind now.

Reply:
> Is it possible that the block device was not in `m_activeBlockDevices` but it is in `m_blockDevices`? And if so should it be removed if that happens?

Yes, this removes from both (take is remove as well). If block disappears from the system, it should be removed from both. It can be removed when for example unlocked device mapper device is locked again (dm gets removed) or you physically detach an external storage from the device.

> Also should we care about `m_pendingBlockDevices` here? I think those might be left behind now.

`m_pendingBlockDevices` is relevant only upon the first load when we get each block one by one.

Reply:
Erm. Sorry, I phrased my question a bit poorly. If a block is in `m_blockDevices` but not in `m_activeBlockDevices` then delete block doesn't do anything but should we delete the object that had its pointer removed from `m_blockDevices`? This was the thing I was wondering about.

Reply:
Need to do quite a bit of `git rebase -i` to get these commits organized at the first place. This clearly slipped in on a rename. Good catch! :+1: 

Swapping order is enough as m_blockDevices has all the blocks always. I think that your concern was a bit theoretical as I have not need that causing issues but better to be safe than sorry.